### ISSUE TITLE
Synchronizing highlights.

### ISSUE THREAD

**Issue by jordwalke** [lunixbochs/actualvim#76]
Could there be a way to let Vim perform the filetype detection, and then let Vim perform the syntax highlighting, but then map that back into something that looks decent in Sublime? Imagine a special colorscheme in Vim, whose only job it is to identify various colors to be mapped to equivalent sublime colorscheme colors.

For example, I have spellpad highlighting using Syntastic in Vim and these plugins could be mapped to a red background/underline in Sublime. Also, various syntax highlighters I've defined for Vim could be translated into Sublime this way.

**Comment by lunixbochs** [comment]
I asked about this yesterday. @justinmk says highlighting is only stored to the screen grid in vim, so we can't borrow highlights right now.

**Comment by lunixbochs** [comment]
From chat:

> Ryan Hileman @lunixbochs Mar 14 13:44
> is there any way to get a buffer's highlights?
> 
> Justin M. Keyes @justinmk Mar 14 13:45
> you mean like a map of the syntax groups? Not yet, but it seems like a natural thing to expose in the api
> vim does have things like synstack(), synID(), hlID(), etc, already.
> But i'm not aware of any way to get the full picture
> the problem is that that info is mixed into the screen grid, not a separate structure AFAIK. but it's something i hope we can work on in 2018...
> 
> matrixbot @matrixbot Mar 14 14:01
> raichoo on Freenode Ah nice, I've been hoping for something like this to happen sometime.
> 
> Justin M. Keyes @justinmk Mar 14 14:02
> it's a ways off, unless someone randomly goes nuts with it
> 
> matrixbot @matrixbot Mar 14 14:03
> raichoo on Freenode It's currently way beyond my capabilities otherwise I'd be up for the task. I just started to find my way around the codebase.
> 
> Justin M. Keyes @justinmk Mar 14 14:03
> the amazing thing about that is it would allow a child nvim instance to asynchronously calculate syntax info then transfer it to the parent
> 
> matrixbot @matrixbot Mar 14 14:03
> raichoo on Freenode That would be great.
> 
> Justin M. Keyes @justinmk Mar 14 14:04
> or any client that conforms to the map structure, could caluclate however they want
> 
> matrixbot @matrixbot Mar 14 14:06
> raichoo on Freenode Would also be nice to gather some semantic information so one could have things like "semantic-objects" to do things like "delete inner pattern match branch" etc. Not sure if I'm making sense.
> 
> Justin M. Keyes @justinmk Mar 14 14:07
> oh, i asked for that on vim_dev. it's actually on brams' todo.txxt

**Comment by justinmk** [comment]
If you only want highlights (not syntax semantics), those are already sent as UI events, see `:help rpc-remote-ui` and the "highlight_set" event. That might avoid screen scraping.

**Comment by lunixbochs** [comment]
That's what I meant by screen scraping. `highlight_set` only affects characters drawn to the Neovim screen. To show them in Sublime I need to translate highlights from the virtual Neovim screen to the actual text, which is kinda a hack. I am currently syncing viewport size to Neovim, but not scroll position as Sublime can scroll *very* fast.

**Comment by jordwalke** [comment]
The primary highlights that I think are most important are:

- Incremental search and search results when navigating via `n/N` or searching with `/`.
- Underlined (spellbad) for syntax checkers.

The syntax *semantics* are not really important. I know vim understands them, and to the user, those are just a means of calculating the colors in the grid. So it sounds like screen scraping will work - but how much of a hack is it?

> I need to translate highlights from the virtual Neovim screen to the actual text, which is kinda a hack.

How involved is that? I can't imagine it is harder than whatever you did to synchronize scrolling between the two worlds. Wouldn't it be okay if the syntax near the viewport was highlighted and then content offscreen would be colored as information becomes available? If you scroll further than you have color information, I imagine it would be okay to show gray text, and then even fade the colors in as they become available (I wish it could be done via an animation to avoid the jaring FOUC (flash of unstyled content)). Asynchronously highlighting colors isn't so bad, but text switching from normal to italic asynchronously is really terrible imho.

**Comment by lunixbochs** [comment]
Scrolling isn't actually synchronized. There's just a check to make sure the cursor is visible in Sublime. But that's on my shortlist.

I will sync basic highlights relative to the cursor. That's pretty much guaranteed to work for search as it moves the cursor.

**Comment by jordwalke** [comment]
My incremental search might move the cursor but it highlights many things in the buffer as I type. Will that work?

<img width="546" alt="screen shot 2017-03-17 at 4 49 11 am" src="https://cloud.githubusercontent.com/assets/977348/24041983/26e97e92-0acd-11e7-970e-ae869c4deba5.png">

**Comment by lunixbochs** [comment]
I should be able to detect any highlights on the same screen (vim, not sublime, though they are set to the same size) as the cursor. This means it should only miss highlights if you use Sublime's navigation (mouse scroll, etc) to move the screen away from the cursor.

**Comment by jordwalke** [comment]
That seems totally reasonable. Note that I use this amazing plugin called inc-search, and what it does, is it allows me to search incrementally using `/` but then hit `tab`/`shift+tab` to move the cursor between the highlight results, without exiting incremental search. You can jump through the whole file while incrementally searching. This would still work with your approach right since it uses vim's own internal jumping around?

Basically, is it correct to say, that if you only ever use vim's own navigation, and not sublime's scrolling, the screen will always have accurate highlighting? I would be very excited to just have that.

**Comment by lunixbochs** [comment]
Sounds about right. And here "highlight" means when vim displays a background color, I will try to apply a background color in sublime as well. I'm not planning to parse foreground colors, and honestly I should probably make sure neovim's syntax highlighting is off.

**Comment by jordwalke** [comment]
Why not parse foreground color though if you're already doing background? Is your goal only to see where search highlighting is? There's some other valuable ones as well such as "SpellBad" which is used for error highlighting - where those error markers are determined by VimScripts.
Is it much harder just to parse all colors background and foreground?

My thinking was that you can inject a special color scheme into Vim, that lets you perfectly disambiguate:

- `IncSearch` : `bg=#000000, fg=#000001`
- `Constant`: `bg=#000002, fg=#000003`
- ... and so on


Which is a *terrible* looking color scheme, but allows you to then map these colors to their equivalent Sublime color theme because you always know that `#000002` means the bg color for "Constant".  No hex color is repeated among syntax elements - that is important. You would inject this `sublimeMappingColorScheme.vim` into the vim instance. If fact, other `neovim` frontends could use this to perform their own reconstruction.

So with this, all your vim language plugins work and highlight their grammar elements correctly, yet you get to enjoy a beautiful Sublime color scheme that matches regular Sublime.

**Comment by lunixbochs** [comment]
I don't think you can change foreground text colors in Sublime from a plugin without resorting to massive hacks: https://github.com/SublimeTextIssues/Core/issues/817

This will also have an update lag as I need to wait for vim to finish drawing the screen, won't work with the minimap, won't work for scrolling... foreground is extremely low priority.

Sublime has spellchecking and pretty good error checking plugins, which will have much better UI integration for now.

**Comment by jordwalke** [comment]
So what about spellbad, which allows underlining of errors highlighted by Vim plugins? How would that play with your current approach?

**Comment by lunixbochs** [comment]
I can do underlines in Sublime.

**Comment by lunixbochs** [comment]
Check out the highlight branch for the initial PoC.
![highlights](https://cloud.githubusercontent.com/assets/117642/24066812/06a8ca78-0b4c-11e7-8cbf-7615d993f4ea.png)

**Comment by jordwalke** [comment]
Whoa. How do you do this? Does it effect performance? :D

**Comment by lunixbochs** [comment]
Screen scraping neovim, dumping window attributes and correlating with the underlying text. Probably affects perf because I get a ton of redraw events for every action, but it wasn't *that* slow and it should scale okay. Can optimize a ton still (mostly by making it apply a few dozen less times a second).

There's several moderate bugs still, and it assumes all special character attributes are a background highlight. I think even syntax highlights. Which explains why some files had a huge highlighted section. Probably where the vim screen was centered.

**Comment by lunixbochs** [comment]
Basic highlight support pushed to master.
